src/profile-logic/zip-files.js

```diff
@@ -0,0 +1,269 @@
+// @flow
+
+import JSZip, { type JSZipFile } from 'jszip';
+import { ensureIsValidTabSlug } from '../utils/flow';
+export type IndexIntoZipFileTable = number;
+
+/**
+ * The zip file table takes the files data structure of {[filePath]: fileContents} and
+ * maps it into a hierarchical table that can be used by the TreeView component to
+ * generate a file tree.
+ */
+export type ZipFileTable = {|
+  prefix: Array<IndexIntoZipFileTable | null>,
+  path: string[], // e.g. "profile_tresize/tresize/cycle_0.profile"
+  partName: string[], // e.g. "cycle_0.profile" or "tresize"
+  file: Array<JSZipFile | null>,
+  depth: number[],
+  length: number,
+|};
+
+export type ZipDisplayData = {|
+  +name: string,
+  +url: null | string,
+  +zipTableIndex: IndexIntoZipFileTable,
+|};
+
+export function createZipTable(zipEntries: JSZip): ZipFileTable {
+  const fullPaths = [];
+  for (const fileName in zipEntries.files) {
+    if (zipEntries.files.hasOwnProperty(fileName)) {
+      const file = zipEntries.files[fileName];
+      if (!file.dir) {
+        fullPaths.push(fileName);
+      }
+    }
+  }
+  const pathToFilesTableIndex: Map<string, IndexIntoZipFileTable> = new Map();
+  const filesTable: ZipFileTable = {
+    prefix: [],
+    path: [],
+    partName: [],
+    file: [],
+    depth: [],
+    length: 0,
+  };
+
+  for (let i = 0; i < fullPaths.length; i++) {
+    // e.g.: 'profile_tresize/tresize/cycle_0.profile'
+    const fullPath = fullPaths[i];
+    // e.g.: ['profile_tresize', 'tresize', 'cycle_0.profile']
+    const pathParts = fullPath.split('/').filter(part => part);
+
+    let path = '';
+    let prefixIndex = null;
+    for (let j = 0; j < pathParts.length; j++) {
+      // Go through each path part to assemble the table
+      const pathPart = pathParts[j];
+
+      // Add the path part to the path.
+      if (path) {
+        path += '/' + pathPart;
+      } else {
+        path = pathPart;
+      }
+
+      // This part of the path may already exist.
+      const existingIndex = pathToFilesTableIndex.get(path);
+      if (existingIndex !== undefined) {
+        // This folder was already added, so skip it, but remember the prefix.
+        prefixIndex = existingIndex;
+        continue;
+      }
+
+      const index = filesTable.length++;
+      filesTable.prefix[index] = prefixIndex;
+      filesTable.path[index] = path;
+      filesTable.partName[index] = pathPart;
+      filesTable.depth[index] = j;
+      filesTable.file[index] =
+        j + 1 === pathParts.length ? zipEntries.files[fullPath] : null;
+      pathToFilesTableIndex.set(path, index);
+      // Remember this index as the prefix.
+      prefixIndex = index;
+    }
+  }
+  return filesTable;
+}
+
+export function getZipFileMaxDepth(zipFileTable: ZipFileTable | null): number {
+  if (!zipFileTable) {
+    return 0;
+  }
+  let maxDepth = 0;
+  for (let i = 0; i < zipFileTable.length; i++) {
+    maxDepth = Math.max(maxDepth, zipFileTable.depth[i]);
+  }
```

Comment:
Maybe `Math.max(0, ...zipFileTable.depth)` ?

Reply:
But I wonder if that's not something that could be in ZipFileTable and just filled in at parse time ?

Reply:
I'd prefer C-like code here, and the function follows other patterns for the other trees, so I'm going to keep it.

Reply:
`Math.max(0, ...zipFileTable.depth)` replaces the whole `for` loop, I think it's faster to do it in one call ([where the loop is done in C++](https://searchfox.org/mozilla-central/rev/769222fadff46164f8cc0dc7a0bae5a60dc2f335/js/src/jsmath.cpp#595)) rather than looping in JS.